payments_service/braintree/tests/test_forms.py

```diff
@@ -36,6 +40,124 @@ def submit(self, expect_errors=False, data=None, overrides=None,
+        getter.side_effect = ObjectDoesNotExist
+
+    def expect_existing_buyer(self, uuid='some-uuid',
+                              authenticated=False):
+        self.solitude.generic.buyer.get_object_or_404.return_value = {
+            'resource_pk': 1244,
+            'uuid': uuid,
+            'authenticated': authenticated,
+        }
+
+    def test_too_many_pay_methods(self):
+        form = self.submit(
+            expect_errors=True,
+            overrides=dict(pay_method_uri='/my/saved/paymethod',
+                           pay_method_nonce='some-nonce'),
+        )
+        assert '__all__' in form.errors, (
+            form.errors.as_text()
+        )
+
+    def test_missing_pay_method(self):
+        form = self.submit(
+            expect_errors=True,
+            overrides=dict(pay_method_uri=None,
+                           pay_method_nonce=None),
+        )
+        assert '__all__' in form.errors, (
+            form.errors.as_text()
+        )
+
+    def test_email_only_recurring_donation_creates_user(self):
+        self.expect_non_existant_buyer()
+        email = 'someone@somewhere.org'
+        self.solitude.generic.buyer.post.return_value = {
+            'uuid': 'created-uuid',
+            'resource_pk': 1234,
+        }
+
+        form = self.submit(
+            user=False,
+            overrides=dict(plan_id='org-recurring-donation',
+                           email=email))
+
+        assert self.solitude.generic.buyer.post.called
+        args = self.solitude.generic.buyer.post.call_args[0][0]
+        eq_(args['email'], email)
+        eq_(args['authenticated'], False)
+        assert args['uuid'].startswith('service:unauthenticated:'), (
+            args['uuid']
+        )
+        eq_(form.user.uuid, 'created-uuid')
+
+    def test_reuse_email_buyer_if_not_authenticated(self):
+        self.expect_existing_buyer(uuid='existing-uuid',
+                                   authenticated=False)
+        email = 'someone@somewhere.org'
+        form = self.submit(
+            user=False,
+            overrides=dict(plan_id='org-recurring-donation',
+                           email=email))
+        eq_(form.user.uuid, 'existing-uuid')
+        assert not self.solitude.generic.buyer.post.called
+
+    def test_cannot_reuse_email_buyer_if_they_were_authenticated(self):
+        self.expect_existing_buyer(authenticated=True)
+        form = self.submit(
+            expect_errors=True,
+            user=False,
+            overrides=dict(plan_id='org-recurring-donation',
+                           email='someone@somewhere.org'))
+        assert '__all__' in form.errors, (
+            form.errors.as_text()
+        )
+
+    def test_recurring_donation_requires_email(self):
+        form = self.submit(
+            expect_errors=True,
+            user=False,
+            overrides=dict(plan_id='org-recurring-donation'))
+        assert 'plan_id' in form.errors, (
```

Comment:
I tend to shy away from testing Django validators directly because they have tests of their own: https://github.com/django/django/blob/master/tests/forms_tests/tests/test_fields.py#L744